### ISSUE TITLE
Problem with formatter: 'Request textDocument/formatting failed.   Message: [NoMethodError] undefined method `+' for nil:NilClass'

### ISSUE THREAD

**Issue by electr0sheep** [castwide/solargraph#142]
The problem I believe is this line:
https://github.com/castwide/solargraph/blob/master/lib/solargraph/language_server/message/text_document/formatting.rb#L17

I am trying to determine what 'lines' is when I run this and will update if I find out

**Comment by castwide** [comment]
My first guess would be that the formatting fails for some reason, such as a syntax error in the file, so the output is empty. The lack of error checking--at least checking the status code of the process--is a definite problem.

The Open3 call to the `rubocop` executable should probably be replaced with an inline call to a `RuboCop::Runner`, similar to how the [RuboCop diagnostics reporter](https://github.com/castwide/solargraph/blob/master/lib/solargraph/diagnostics/rubocop.rb) does it.

**Comment by electr0sheep** [comment]
I don't believe it is caused by a syntax error. If I run rubocop over the file, it doesn't detect any offenses.

**Comment by castwide** [comment]
Syntax error was just an example. It looks to me like a fatal error from the `rubocop` executable, and it might have nothing to do with the file itself. The base problem I see is that the `Formatting` class doesn't properly report errors, regardless of the root cause.

**Comment by jankeromnes** [comment]
I think I'm hitting this error too:

```
$ gem install solargraph
Fetching parser-2.6.0.0.gem
Fetching ast-2.4.0.gem
Fetching kramdown-1.17.0.gem
Fetching mini_portile2-2.4.0.gem
Fetching htmlentities-4.3.4.gem
Fetching backport-0.3.0.gem
Fetching jaro_winkler-1.5.2.gem
Fetching nokogiri-1.10.1.gem
Fetching reverse_markdown-1.1.0.gem
Fetching parallel-1.13.0.gem
Fetching powerpack-0.1.2.gem
Fetching rainbow-3.0.0.gem
Fetching ruby-progressbar-1.10.0.gem
Fetching unicode-display_width-1.4.1.gem
Fetching rubocop-0.63.1.gem
Fetching thor-0.20.3.gem
Fetching solargraph-0.31.1.gem
Fetching tilt-2.0.9.gem
Fetching yard-0.9.18.gem
Successfully installed backport-0.3.0
Successfully installed htmlentities-4.3.4
Building native extensions. This could take a while...
Successfully installed jaro_winkler-1.5.2
Successfully installed kramdown-1.17.0
Successfully installed ast-2.4.0
Successfully installed parser-2.6.0.0
Successfully installed mini_portile2-2.4.0
Building native extensions. This could take a while...
Successfully installed nokogiri-1.10.1
Successfully installed reverse_markdown-1.1.0
Successfully installed parallel-1.13.0
Successfully installed powerpack-0.1.2
Successfully installed rainbow-3.0.0
Successfully installed ruby-progressbar-1.10.0
Successfully installed unicode-display_width-1.4.1
Successfully installed rubocop-0.63.1
Successfully installed thor-0.20.3
Successfully installed tilt-2.0.9
--------------------------------------------------------------------------------
As of YARD v0.9.2:

RubyGems "--document=yri,yard" hooks are now supported. You can auto-configure
YARD to automatically build the yri index for installed gems by typing:

    $ yard config --gem-install-yri

See `yard config --help` for more information on RubyGems install hooks.

You can also add the following to your .gemspec to have YARD document your gem
on install:

    spec.metadata["yard.run"] = "yri" # use "yard" to build full HTML docs.

--------------------------------------------------------------------------------
Successfully installed yard-0.9.18
Successfully installed solargraph-0.31.1
Parsing documentation for backport-0.3.0
Installing ri documentation for backport-0.3.0
Parsing documentation for htmlentities-4.3.4
Installing ri documentation for htmlentities-4.3.4
Parsing documentation for jaro_winkler-1.5.2
Installing ri documentation for jaro_winkler-1.5.2
Parsing documentation for kramdown-1.17.0
Installing ri documentation for kramdown-1.17.0
Parsing documentation for ast-2.4.0
Installing ri documentation for ast-2.4.0
Parsing documentation for parser-2.6.0.0
Installing ri documentation for parser-2.6.0.0
Parsing documentation for mini_portile2-2.4.0
Installing ri documentation for mini_portile2-2.4.0
Parsing documentation for nokogiri-1.10.1
Installing ri documentation for nokogiri-1.10.1
Parsing documentation for reverse_markdown-1.1.0
Installing ri documentation for reverse_markdown-1.1.0
Parsing documentation for parallel-1.13.0
Installing ri documentation for parallel-1.13.0
Parsing documentation for powerpack-0.1.2
Installing ri documentation for powerpack-0.1.2
Parsing documentation for rainbow-3.0.0
Installing ri documentation for rainbow-3.0.0
Parsing documentation for ruby-progressbar-1.10.0
Installing ri documentation for ruby-progressbar-1.10.0
Parsing documentation for unicode-display_width-1.4.1
Installing ri documentation for unicode-display_width-1.4.1
Parsing documentation for rubocop-0.63.1
Installing ri documentation for rubocop-0.63.1
Parsing documentation for thor-0.20.3
Installing ri documentation for thor-0.20.3
Parsing documentation for tilt-2.0.9
Installing ri documentation for tilt-2.0.9
Parsing documentation for yard-0.9.18
Before reporting this, could you check that the file you're documenting
has proper syntax:

  /home/gitpod/.rvm/rubies/ruby-2.6.0/bin/ruby -c lib/yard/i18n/pot_generator.rb

RDoc is not a full Ruby parser and will fail when fed invalid ruby programs.

The internal error was:

        (NoMethodError) undefined method `[]' for nil:NilClass

ERROR:  While executing gem ... (NoMethodError)
    undefined method `[]' for nil:NilClass
```

Here is a snapshot environment that reproduces this bug:

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io#snapshot/3ff5ad68-0ef4-4df3-b47f-7105f98488b3)

**Comment by MarcusHoile** [comment]
I get the same error. When i run rubocop on the file there is no `=======` in the output so the index will return nil on this line https://github.com/castwide/solargraph/blob/master/lib/solargraph/language_server/message/text_document/formatting.rb#L16

**Comment by gencer** [comment]
+1 for this.

**Comment by castwide** [comment]
I still haven't been able to reproduce this issue. Two questions for anyone who can:

1. What version of RuboCop is installed?
2. Can you provide example code that triggers the error?

**Comment by gencer** [comment]
```
$ rubocop -v
0.71.0
```

Sample code:

```
# frozen_string_literal: true

class Chat::OrganizationMember < Chat::Model(:organization_members)
  self.raise_on_save_failure = false

  many_to_one :organization

  STATUS_ONLINE = 1.freeze
  STATUS_OFFLINE = 2.freeze
  STATUS_AWAY = 3.freeze

  def current_status
    return 2 unless status
    unless status_on.nil?
      diff = (Time.now.utc - status_on) / 1.minutes
      return 2 if diff > 10
    end
    return status
  end

end
```

**Comment by gencer** [comment]
I think this is happening because of .rubocop.yml file.

Please test with:

```
require: rubocop-rails
AllCops:
  # If you want the rule only apply to a specific folder/file
  Include:
    - '**/Rakefile'
    - '**/config.ru'
    - 'app/**/*'
  # If you want the rule not to apply to a specific folder/file
  Exclude:
    - 'db/**/*'
    - 'config/**/*'
    - 'script/**/*'
    - 'Gemfile'
    - '/Gemfile'
    - 'gems.rb'
    - 'Gemfile.lock'
    - 'gems.locked'
#  DisplayCopNames: true
#  Rails:
#    Enabled: true
#    - !ruby/regexp /old_and_unused\.rb$/
Layout/TrailingWhitespace:
  Enabled: false
Metrics/ParameterLists:
  Enabled: false
Metrics/LineLength:
  # This will disable the rule completely, regardless what other options you put
  Enabled: false
  # Change the default 80 chars limit value
  Max: 250
Metrics/MethodLength:
  Enabled: false
Metrics/BlockLength:
  Enabled: false
Metrics/BlockNesting:
  Enabled: false
# Style/EndOfLine:
# EnforcedStyle: lf
ClassLength:
  Max: 1000 # or whatever ends up being appropriate
# Checks the indentation of the first key in a hash literal.
Layout/IndentHash:
  # The value `special_inside_parentheses` means that hash literals with braces
  # that have their opening brace on the same line as a surrounding opening
  # round parenthesis, shall have their first key indented relative to the
  # first position inside the parenthesis.
  #
  # The value `consistent` means that the indentation of the first key shall
  # always be relative to the first position of the line where the opening
  # brace is.
  #
  # The value `align_braces` means that the indentation of the first key shall
  # always be relative to the position of the opening brace.
  EnforcedStyle: special_inside_parentheses
  SupportedStyles:
    - special_inside_parentheses
    - consistent
    - align_braces
  # By default, the indentation width from `Layout/IndentationWidth` is used
  # But it can be overridden by setting this parameter
  IndentationWidth: ~
# Layout/IndentationConsistency:
#   # The difference between `rails` and `normal` is that the `rails` style
#   # prescribes that in classes and modules the `protected` and `private`
#   # modifier keywords shall be indented the same as public methods and that
#   # protected and private members shall be indented one step more than the
#   # modifiers. Other than that, both styles mean that entities on the same
#   # logical depth shall have the same indentation.
#   EnforcedStyle: normal
#   # AllowProtectedAndPrivateMethodsBeIndented: true
#   SupportedStyles:
#   #    - normal
#   - rails
Layout/IndentationWidth:
  # Number of spaces for each indentation level.
  Width: 2
  IgnoredPatterns: []
Lint/PercentStringArray:
  Enabled: false
Style/PercentLiteralDelimiters:
  Enabled: false
Style/PreferredHashMethods:
  Enabled: false
#  PreferredDelimiters:
#    "%i": '()'
Bundler/OrderedGems:
  Enabled: false
Security/YAMLLoad:
  Enabled: false
```

**Comment by castwide** [comment]
Yep, that did it for me. Thanks, @gencer! I'll see if I can track down the error from the spawned process.

**Comment by castwide** [comment]
Here's the error:

```
The `Layout/IndentHash` cop has been renamed to `Layout/IndentFirstHashElement`
```

I'll update the gem so it sends the original error from the RuboCop process instead of the subsequent error from failing to parse an empty result.

**Comment by gencer** [comment]
I've updated that setting according to recommendation from rubocop. However, it still produces an error.

**Comment by castwide** [comment]
Updating that setting worked for me. There might be some other difference in our default RuboCop configurations or some other environment setting.

I pushed a change to the master branch that reports the specific RuboCop error. It'll be in the next gem release.

**Comment by castwide** [comment]
The fix is released in 0.34.0. You'll probably still get an error, but at least now it should tell you exactly what caused it.

**Comment by gencer** [comment]
Update: Confirmed, This fixed my issue.
